rapthor/lib/field.py

```diff
@@ -976,16 +981,69 @@ def makeWCS(self):
         w.wcs.set_pv([(2, 1, 45.0)])
         self.wcs = w
 
-    def check_selfcal_convergence(self):
+    def check_selfcal_progress(self, convergence_ratio=None, divergence_ratio=None):
         """
-        Checks whether selfcal has converged or not on a sector-by-sector basis
+        Checks whether selfcal has converged or diverged by comparing the current
+        image noise to that of the previous cycle
+
+        Parameters
+        ----------
+        convergence_ratio : float, optional
+            The minimum ratio of the current noise to the previous noise above
+            which selfcal is considered to have converged (must be in the range
+            0.5 -- 2). E.g., convergence_ratio = 0.95 means that the image noise
+            must improve by ~ 5% or more from the previous cycle for selfcal to be
+            considered as nonconverged
+        divergence_ratio : float, optional
+            The minimum ratio of the current noise to the previous noise above
+            which selfcal is considered to have diverged (must be > 1). E.g.,
+            divergence_ratio = 1.1 means that, if image noise worsens by ~ 10%
+            or more from the previous cycle, selfcal is considered to have
+            diverged
 
         Returns
         -------
-        result : bool
-            True if all sectors have converged, False if not
+        converged : bool
+            True if selfcal has converged, False if not
+        diverged : bool
+            True if selfcal has diverged, False if not
         """
-        return False
+        if convergence_ratio is None:
+            convergence_ratio = self.convergence_ratio
+        if divergence_ratio is None:
+            divergence_ratio = self.divergence_ratio
+
+        # Check that convergence and divergence limits are sensible
+        if convergence_ratio > 2.0:
+            convergence_ratio = 2.0
+        if convergence_ratio < 0.5:
+            convergence_ratio = 0.5
+        if divergence_ratio < 1.0:
+            divergence_ratio = 1.0
+
+        if self.field_image_filename_prev is None:
+            # No previous iteration, so report nonconvergence
+            return False
+
+        # Get noise from previous and current images
+        image = FITSImage(self.field_image_filename_prev)
+        image.calc_noise()
+        rmspre = image.noise
+        image = FITSImage(self.field_image_filename)
+        image.calc_noise()
+        rmspost = image.noise
```

Comment:
It this an expensive operation? Otherwise, it might be good to calculate the noise while you still have the image in memory.

Reply:
It is somewhat expensive, but still only a few minutes I think. If it turns out to be much longer, we should indeed revisit this.